lib/ecto/query/api.ex

```diff
@@ -272,6 +272,39 @@ defmodule Ecto.Query.API do
   def fragment(fragments), do: doc! [fragments]
 
   @doc """
+  Send an unsafe fragment directly to the database.
+
+  WARNING: sending an unsafe fragment may expose your application to sql
+  injection attacks. Think twice before using it and try to reach the same
+  result using the safe fragment/1 API.
+
+  Sometimes it is not possible to represent all possible databases queries
+  using Ecto's syntax or fragment/1, expecially when dealing with specific
+  database construct and dynamic fields.
+
+  To overcome this, it is possible to build the expression as a binary
+  outside the query expression and send it to the database by interpolating
+  the value in the query.
+
+      def window_sorting(field, dir) do
+        expr = "ROW_NUMBER() OVER(ORDER BY \#{field} \#{dir})"
```

Comment:
This is not a good example as I already do the same via normal macro's.  Perhaps show how to do a multi-prefix join or a UNION or so?

Reply:
mmmh I don't think is possible with normal macros when the fields are dynamic, or at least I was not able to do it.

For example, my use case:
```
  defp add_query_ordering_with_cursor(model, sorters) do
    sort = sorters
    |> Enum.map(fn %{sort: key, direction: dir} ->
      key = case key do
        k when is_atom(k) -> Atom.to_string(k)
        k when is_binary(k) -> k
      end

      dir = case dir do
        :asc -> "ASC"
        :desc -> "DESC"
      end

      "#{key} #{dir}"
    end)
    |> Enum.join(",")

    sort = case sort do
      "" -> "id"
      _ -> sort
    end

    fragment = "ROW_NUMBER() over(ORDER BY #{sort})"
    from m in model,
      select_merge: %{m |
        id: unsafe_fragment(^fragment)
      }

  end
```
and after that I add other selects, filters and limits.
Is really possible with standard macros?

Reply:
> Is really possible with standard macros?

Not macro's, it's a built-in function in Ecto.Query:  https://hexdocs.pm/ecto/Ecto.Query.API.html#field/2  :-)

Reply:
yes I know it, but if you look how ROW_NUMBER works, it must be into the select clause and it has a special syntax with `OVER (order_by_clause)`  where the order_by_clause is *not* the order_by of the Ecto syntax.

an example query:
`SELECT *, ROW_NUMBER() OVER (ORDER by realm DESC) FROM "domains"`

how is rendered with standard Ecto DSL ? I've bashed by head for a couple of days and from what I've read around is just not possible.

Reply:
Here's an actual example from my source:
```elixir
              select: %{
                _rank: fragment("rank() OVER (PARTITION BY ?, ? ORDER BY ? DESC)",
                  crse.scbcrse_crse_numb, crse.scbcrse_subj_code, field(crse, sorted_field)),
...
```
Is that not what like you are wanting to do (you can of course use `field` on any/all of the fragment inputs too)?

Reply:
my problem is that the number of fields is not fixed nor are the directions.
My order clause may be from one to N (where N are all schema fields), and each one being sorted in a different direction.